python/revng/tupletree/__init__.py

```diff
@@ -153,7 +155,7 @@ def __post_init__(self):
 
     def __setattr__(self, key, value):
         # Prevent setting undefined attributes
-        if self.__dataclass_fields__.get(key) is None:
+        if self.__dataclass_fields__.get(key) is None and key != "project":
```

Comment:
If I check for undefined attr with `if key not in dir(self):` would this be acceptable? The issue with the current implementation is that it only checks the attrs in the `dataclass` with `dir(self)` it will check other inherited classes too.

Reply:
Use `dataclasses.fields`, like:
```python
if key not in [f.name for f in fields(self)]:
```

Reply:
Ok, let's then do:
```python
if key not in [f.name for f in fields(self)] or key.startswith("_"):
```

Since `_project` is private